temoa/utilities/db_migration_v3_1_to_v4.py

```
@@ -0,0 +1,290 @@
+        return 0
+    # filter out rows that are entirely NULL
+    filtered = [r for r in rows if any(v is not None for v in r)]
+    if not filtered:
+        return 0
+    placeholders = ','.join(['?'] * len(insert_new_cols))
+    q = f'INSERT OR REPLACE INTO {new_table} ({",".join(insert_new_cols)}) VALUES ({placeholders})'
+    con_new.executemany(q, filtered)
+    return len(filtered)
+
+
+def migrate_all(args: argparse.Namespace) -> None:
```

Ruff (0.15.9)

[warning] Too many branches (18 > 12)

(PLR0912)